### PR TITLE
add focus tactic

### PR DESCRIPTION
**Pull request by bodeveix**
Add a focus tactic that allows changing the currently focused goal.

### CODE REVIEW

File: src/handle/tactic.ml [hunk 1]
```diff
@@ -340,6 +342,18 @@ let p_rwpatt_of_string (pos:popt) (t:term): p_rwpatt option =
            Some (Parsing.Parser.Lp.parse_rwpatt_string p string)
   | _ -> fatal pos "not a string literal"
 
+(** [p_int_of_string pos t] returns the int contained in a string literal
+    term [t]. *)
+let p_int_of_string : popt -> term -> int = fun pos t ->
+  match t with
+  | Symb s when String.is_string_literal s.sym_name ->
+      begin
+        let string = String.trim (remove_quotes s.sym_name) in
```
Comment (fblanqui) [inline src/handle/tactic.ml]: Why using String.trim? s.sym_name cannot contain spaces.

File: src/handle/tactic.ml [hunk 2]
```diff
@@ -340,6 +342,18 @@ let p_rwpatt_of_string (pos:popt) (t:term): p_rwpatt option =
...
+        try int_of_string string
+        with _ -> fatal pos "string should contain an int"
```
Comment (fblanqui) [inline src/handle/tactic.ml]: explicit exception name like in https://github.com/Deducteam/lambdapi/blob/f3286759f3b8a1ec3aaa55706a67dd36eb020de5/src/handle/query.ml#L92-L94

File: src/handle/tactic.ml [hunk 3]
```diff
@@ -340,6 +342,18 @@ let p_rwpatt_of_string (pos:popt) (t:term): p_rwpatt option =
            Some (Parsing.Parser.Lp.parse_rwpatt_string p string)
   | _ -> fatal pos "not a string literal"
 
+(** [p_int_of_string pos t] returns the int contained in a string literal
+    term [t]. *)
+let p_int_of_string : popt -> term -> int = fun pos t ->
```
Comment (fblanqui) [inline src/handle/tactic.ml]: rename into int_of_term

File: src/handle/tactic.ml [hunk 4]
```diff
@@ -512,6 +528,18 @@ let handle (ss:Sig_state.t) (sym_pos:popt) (priv:bool)
       let id =  Pos.make pos (P_Abst(vparam,idbody)) in
       let pt = Pos.make pos (P_Appl(id,Pos.make pos P_Wild)) in
       tac_refine pos ps gt gs (new_problem()) (scope pt)
+  | P_tac_focus n -> begin
```
Comment (fblanqui) [inline src/handle/tactic.ml]: I think that this should be moved after https://github.com/Deducteam/lambdapi/blob/f3286759f3b8a1ec3aaa55706a67dd36eb020de5/src/handle/tactic.ml#L455 .

File: src/handle/tactic.ml [hunk 5]
```diff
@@ -512,6 +528,18 @@ let handle (ss:Sig_state.t) (sym_pos:popt) (priv:bool)
...
+      let rec extract n l =
+        match l with
+        | [] -> raise Not_found
+        | x::l' -> if n = 0 then x,l'
+                   else let (y,r) = extract (n-1) l' in (y,x::r) in
+      try
+        let (x,gs) = extract n ps.proof_goals in
+        {ps with proof_goals = x::gs}
+      with
+        Not_found -> ps
```
Comment (fblanqui) [inline src/handle/tactic.ml]: a fatal should be raised if the focus is out of bound

File: editors/emacs/lambdapi-smie.el [hunk 6]
```diff
@@ -102,7 +102,7 @@ Indent by `lambdapi-indent-basic' in proofs, and 0 otherwise."
               ("change" sterm)
               ("eval" sterm)
               ("fail")
-              ("focus" ident)
+              ("focus" int)
```
Comment (fblanqui) [inline editors/emacs/lambdapi-smie.el]: I don't think that int is defined, is it? just leave ident otherwise

File: doc/tactics.rst [hunk 7]
```diff
@@ -40,6 +40,14 @@ Proves the current goal if it is (an instance of) an hypothesis.
 
 .. _fail:
 
+
+.. _focus:
+
+``focus <n>``
+-------------
+
+Makes the nth goal become the current goal.
```
Comment (fblanqui) [inline doc/tactics.rst]: nth -> n-th + add " (n≥2)" at the end (focus 1 is useless)

File: doc/lambdapi.bnf [hunk 8]
```diff
@@ -101,6 +101,7 @@
            | "change" <term>
            | "eval" <term>
            | "fail"
+           | "focus" <int>
```
Comment (fblanqui) [inline doc/lambdapi.bnf]: `<integer>`

File: src/parsing/lpParser.ml [hunk 9]
```diff
@@ -1014,6 +1017,16 @@ and tactic (lb:lexbuf): p_tactic =
...
       consume_token lb;
       make_pos pos1 P_tac_fail
+  | FOCUS ->
+      let pos1 = current_pos() in
+      consume_token lb;
+      begin
+        match current_token() with
+        | INT n ->
+            consume_token lb;
+            extend_pos (*__FUNCTION__*) pos1 (P_tac_focus(int_of_string n))
```
Comment (fblanqui) [inline src/parsing/lpParser.ml]: P_tac_focus should take the string n as argument. The conversion to int should be done in tactic.ml.

File: src/parsing/pretty.ml [hunk 10]
```diff
@@ -315,6 +316,7 @@ let rec tactic : p_tactic pp = fun ppf { elt;  _ } ->
   | P_tac_change t -> out ppf "change %a" term t
   | P_tac_eval t -> out ppf "eval %a" term t
   | P_tac_fail -> out ppf "fail"
+  | P_tac_focus n -> out ppf "focus %d" n
```
Comment (fblanqui) [inline src/parsing/pretty.ml]: %s

File: src/parsing/syntax.ml [hunk 11]
```diff
@@ -272,6 +272,7 @@ type p_tactic_aux =
   | P_tac_change of p_term
   | P_tac_eval of p_term
   | P_tac_fail
+  | P_tac_focus of int
```
Comment (fblanqui) [inline src/parsing/syntax.ml]: of string

File: src/handle/tactic.ml [hunk 12]
```diff
@@ -512,6 +528,18 @@ let handle (ss:Sig_state.t) (sym_pos:popt) (priv:bool)
...
+    end
```
Comment (fblanqui) [inline src/handle/tactic.ml]: Can't we code this with a tail-recursive function?
For instance, something like (in pseudo code):
```
let move_nth (n:int) (l:'a list): 'a list =
  let rec move (acc:'a list) n l =
    match l with
     | [] -> invalid_arg "remove_nth"
     | x::l -> if n <= 0 then x :: L.rev_append acc l else move (x:acc) (n-1) l
  in move []
```
to be added in lplib/list.ml
then the tactic code would be something like:
```
  let n = try ... int_of_string ... in
  if n < 2 || n > 1 + List.length gs then fatal pos ...
  else {ps with proof_goals = move_nth (n-1) (g::gs)}
```

**Review by fblanqui — CHANGES_REQUESTED** [review]
Thanks JP for your PR. Here are some comments:

File: src/handle/tactic.ml (hunk 1 above)
Comment (bodeveix) [reply src/handle/tactic.ml]: If #focus is used inside eval, it is declared as having the type String -> Tactic. Managing an Int would be too complex.

File: src/handle/tactic.ml (hunk 2 above)
Comment (bodeveix) [reply src/handle/tactic.ml]: OK I use fatal ...

File: src/handle/tactic.ml (hunk 3 above)
Comment (bodeveix) [reply src/handle/tactic.ml]: OK

File: doc/lambdapi.bnf (hunk 8 above)
Comment (bodeveix) [reply doc/lambdapi.bnf]: OK

File: src/parsing/lpParser.ml [hunk 13]
```diff
@@ -1014,6 +1017,16 @@ and tactic (lb:lexbuf): p_tactic =
...
       make_pos pos1 P_tac_fail
+  | FOCUS ->
+      let pos1 = current_pos() in
+      consume_token lb;
+      begin
+        match current_token() with
+        | INT n ->
+            consume_token lb;
+            extend_pos (*__FUNCTION__*) pos1 (P_tac_focus n)
+        | _ -> expected "" [INT "<int>"]
```
Comment (fblanqui) [inline src/parsing/lpParser.ml]: should be `INT ""` (the argument of INT is ignored)

File: src/lplib/list.ml [hunk 14]
```diff
@@ -315,6 +315,19 @@ let pos : ('a -> bool) -> 'a list -> int = fun f ->
...
+let move_nth (n:int) (l:'a list): 'a list =
+  let rec move (acc:'a list) n l =
+    match l with
+     | [] -> invalid_arg "remove_nth"
+     | x::l -> if n <= 0 then x :: rev_append acc l
+               else move (x::acc) (n-1) l
+  in move [] n l
+let _ =
+  let r = move_nth 2 [0;3;2;1;4] in
+  assert (r = [2;0;3;1;4])
```
Comment (fblanqui) [inline src/lplib/list.ml]: please inline r in case we want to add more unit tests
please add a few more unit tests (corner cases when n is the first or the last element)

File: src/lplib/list.ml [hunk 15]
```diff
@@ -315,6 +315,19 @@ let pos : ('a -> bool) -> 'a list -> int = fun f ->
...
+(** [move_nth n l] moves the nth element of l (counting from 0) to its front,
+    or raises [Invalid_argument]. *)
+let move_nth (n:int) (l:'a list): 'a list =
+  let rec move (acc:'a list) n l =
+    match l with
+     | [] -> invalid_arg "remove_nth"
+     | x::l -> if n <= 0 then x :: rev_append acc l
+               else move (x::acc) (n-1) l
+  in move [] n l
+let _ =
```
Comment (fblanqui) [inline src/lplib/list.ml:329]: please add a new and a comment "(* unit tests *)" before

**Review by fblanqui — COMMENTED** [review]
I have a few more change requests.
Please fix also editors/emacs/lambdapi-smie.el.
